packager/mux/mux.go

```diff
@@ -0,0 +1,93 @@
+package mux
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/ampproject/amppackager/packager/certcache"
+	"github.com/ampproject/amppackager/packager/signer"
+	muxp "github.com/ampproject/amppackager/packager/mux/params"
+	"github.com/ampproject/amppackager/packager/util"
+	"github.com/ampproject/amppackager/packager/validitymap"
+)
+
+type mux struct {
+	certCache *certcache.CertCache
+	signer *signer.Signer
+	validityMap *validitymap.ValidityMap
+}
+
+// Implements the HTTP routing strategy used by the amppkg server. The main
+// purpose for rolling our own is to be able to route the following types of
+// URLs:
+//   /priv/doc/https://example.com/esc%61ped%2Furl.html
+// and have the signer sign the URL exactly as it is encoded in the request, in
+// order to meet docs/cache_requirements.md.
+//
+// The default http mux had some problems which led to me investigating 3rd
+// party routers. (TODO(twifkak): Remember and document those problems.)
+//
+// I investigated two 3rd party routers capable of handling catch-all suffix
+// parameters:
+//   https://github.com/julienschmidt/httprouter
+//   https://github.com/dimfeld/httptreemux
+// Both libs unescape their catch-all parameters, making the above use-case
+// impossible. The latter does so despite documenting support for unmodified
+// URL escapings. This, plus a lack of feature needs, led me to believe that
+// writing our own mux was the best approach.
+func New(certCache *certcache.CertCache, signer *signer.Signer, validityMap *validitymap.ValidityMap) http.Handler {
+	return &mux{certCache, signer, validityMap}
+}
+
+func tryTrimPrefix(s, prefix string) (string, bool) {
+	sLen := len(s)
+	trimmed := strings.TrimPrefix(s, prefix)
+	return trimmed, len(trimmed) != sLen
+}
+
+var allowedMethods = map[string]bool{http.MethodGet: true, http.MethodHead: true}
+
+func (this *mux) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	if _, ok := allowedMethods[req.Method]; !ok {
```

Comment:
This only checks for key existence, and not the value. What if the value was false in the map?

Reply:
Done.